src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -674,3 +682,430 @@ def _standard_metadata(self, name: str) -> V1ObjectMeta:
                 "argocd.argoproj.io/sync-options": "Prune=false",
             },
         )
+
+    """It's really tempting to create generic methods for object
+    creation and deletion, that do some fancy exception handling to do
+    retries and ignore error-cases-that-are-normal-operation,
+    especially since Kubernetes methods are regular in both names and
+    argument type and ordering, and we want to do very similar things
+    for many objects.
+
+    How hard could it be, you think, to use getattr() to pluck the
+    right method, and then reuse the guts of creation/retry code?  And
+    then you realize that you could do this to await the creation of
+    an arbitrary Kubernetes object also.
+
+    I am here to warn you, dear reader, that this turns out to be a
+    terrible idea, because it turns errors you could have caught
+    quickly and easily with the type system into runtime errors deep
+    in the FastAPI event loop, with no obvious way to connect the
+    stack trace back to the place you forgot an "await" or misspelled
+    a method name."""
+
+    # Methods for fileserver
+
+    async def create_fileserver(
+        self,
+        username: str,
+        namespace: str,
+        job: V1Job,
+        service: V1Service,
+        gf_ingress: dict[str, Any],
+    ) -> bool:
+        with FILESERVER_LOCK:
```

Comment:
I'm not sure I understand what you're trying to do with this locking. What is being gained by creating all of the objects together atomically (except when there's a Kubernetes failure)?

Reply:
What I am trying to accomplish is that if a deletion is going on, you can't recreate the user at the same time.  Deletion and creation are both pretty fast, so forcing them through a one-at-a-time bottleneck didn't seem so bad.

Reply:
I've reworked the lock a bit but I still don't know if it's right.